src/plugins/gs-plugin-legacy-external-apps.c

```diff
@@ -0,0 +1,220 @@
+get_installed_appstream_app (GsPlugin *plugin,
+			     GsApp *app,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	AsApp *as_app;
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	const char *deploy_dir;
+	g_autoptr(AsStore) store = NULL;
+	g_autoptr(FlatpakInstalledRef) ref = NULL;
+	g_autoptr(GFile) appstream_file = NULL;
+	g_autofree char *appstream_xml = NULL;
+	g_autofree char *appstream_path = NULL;
+
+	ref = flatpak_installation_get_installed_ref (priv->installation,
+						      FLATPAK_REF_KIND_APP,
+						      gs_app_get_flatpak_name (app),
+						      gs_app_get_flatpak_arch (app),
+						      gs_app_get_flatpak_branch (app),
+						      cancellable,
+						      error);
+
+	if (!ref)
+		return NULL;
+
+	deploy_dir = flatpak_installed_ref_get_deploy_dir (ref);
+
+	appstream_xml = g_strdup_printf ("%s.appdata.xml",
+					  gs_app_get_flatpak_name (app));
+	appstream_path = g_build_filename (deploy_dir, "files", "share",
+					   "app-info", "xmls", appstream_xml,
+					   NULL);
+	appstream_file = g_file_new_for_path (appstream_path);
+
+	store = as_store_new ();
+	as_store_set_add_flags (store,
+				AS_STORE_ADD_FLAG_USE_UNIQUE_ID |
+				AS_STORE_ADD_FLAG_USE_MERGE_HEURISTIC);
+	if (!as_store_from_file (store, appstream_file, NULL, cancellable,
+				 error))
+		return NULL;
+
+	as_app = as_store_get_app_by_id (store, gs_app_get_id (app));
+	if (!as_app)
+		g_set_error (error, AS_STORE_ERROR, AS_STORE_ERROR_FAILED,
+			     "Failed to get app %s from its own installation "
+			     "AppStream file.", gs_app_get_unique_id (app));
+
+	return g_object_ref (as_app);
+}
+
+static gboolean
+setup_ext_runtime_version (GsPlugin *plugin,
+			   GsApp *app,
+			   GCancellable *cancellable,
+			   GError **error)
+{
+	const char *runtime_version;
+	g_autoptr(GError) local_error = NULL;
+	g_autoptr(AsApp) as_app = NULL;
+
+	if (!app_is_legacy_external_app (app))
+		return TRUE;
+
+	as_app = get_installed_appstream_app (plugin, app, cancellable,
+					      &local_error);
+
+	if (!as_app) {
+		g_warning ("Failed to get AsApp from installed AppStream "
+			   "data of app '%s': %s", gs_app_get_unique_id (app),
+			   local_error->message);
+		g_propagate_error (error, g_steal_pointer (&local_error));
+		return FALSE;
+	}
+
+	/* get the runtime version that is set in the installed AppStream data */
+	runtime_version = as_app_get_metadata_item (as_app,
+						    LEGACY_RUNTIME_MTD_KEY);
+
+	/* we setup the version of the external runtime used by this external
+	 * app so it is later used by "external-apps-cleaner" plugin when
+	 * removing those runtimes; we use a new key and not the one that is
+	 * already set in the metadata so we verify that this key has been
+	 * set by this plugin (and thus, by what was installed) and not by
+	 * the general AppStream data */
+	gs_app_set_metadata (app, LEGACY_RUNTIME_INSTALLED_MTD_KEY, NULL);
+	gs_app_set_metadata (app, LEGACY_RUNTIME_INSTALLED_MTD_KEY,
+			     runtime_version);
```

Comment:
This is because Richard likes these safety checks for some operations. It prevents assigning metadata already assigned.
I disagree with it and have a recurring discussion about it with Richard but for now we have to go with that.